File: Firefly Original/eclipse workspace/Firefly-common/src/animation/base/AEasingFunction1D.cpp
/*
 * EasingFunction1D.cpp
 *
 *  Created on: Mar 11, 2016
 *      Author: d
 */

#include <animation/base/AEasingFunction1D.h>


AEasingFunction1D::~AEasingFunction1D() {
}

void AEasingFunction1D::bindValue(double* valPtr) {
	if(std::find(bindings.begin(), bindings.end(), valPtr) == bindings.end()) {
		bindings.push_back(valPtr);
	}
}

void AEasingFunction1D::update(double time) {
	for(std::vector<double*>::iterator iter = bindings.begin(); iter != bindings.end(); iter++) {
		**iter = easedValue(time);
	}

	if((time - start) / (finish - start) > 1.0) {
		finishedFlag = true;
	} else {
		finishedFlag = false;
	}
}

double AEasingFunction1D::normalizeInput(double input) {
	double normalized = (input - start) / (finish - start);

	return fmax(fmin(normalized, 1.0), 0.0);
}

bool AEasingFunction1D::finished() {
	return finishedFlag;
}

File: Firefly/Firefly Controller/Assets/Scripts/Animation/Base/AEasingFunction1D.cs
using System;
using System.Collections.Generic;

namespace Firefly
{
    // The C++ easing functions bind to a variable by raw pointer — bindValue(double*)
    // then write through it on update. C# has no equivalent, so bindings are setter
    // delegates instead. Same semantics: bind once, call Update(time), the target
    // updates itself. See Port Notes.

    public abstract class AEasingFunction1D
    {
        protected List<Action<double>> bindings = new List<Action<double>>();
        protected bool finishedFlag = false;

        public double start, finish;
        public double easeFrom, easeTo;

        protected AEasingFunction1D(double start, double finish, double easeFrom, double easeTo)
        {
            this.start = start;
            this.finish = finish;
            this.easeFrom = easeFrom;
            this.easeTo = easeTo;
        }

        protected double NormalizeInput(double input)
        {
            double normalized = (input - start) / (finish - start);
            return Math.Max(Math.Min(normalized, 1.0), 0.0);
        }

        public void BindValue(Action<double> setter)
        {
            if (!bindings.Contains(setter)) bindings.Add(setter);
        }

        public void Update(double time)
        {
            foreach (Action<double> setter in bindings) setter(EasedValue(time));

            finishedFlag = (time - start) / (finish - start) > 1.0;
        }

        public bool Finished() { return finishedFlag; }

        public abstract double EasedValue(double input);
    }
}
